### PR TITLE
Authorization policy: support inline multi-values header semantics

### PR DESCRIPTION
**Pull request by edieYoung**
**Please provide a description of this PR:**

Fix issue: https://github.com/istio/istio/issues/39680
This PR adds support in Authorization Policy to recognize inline multi-values header format by using safe regex match.

For example:
given the allowed header { key: x-foo, value: [a, b]} the request with header {key: x-foo value: "a,b,c"} will be allowed after this fix. Before this fix, this is not allowed.

### CODE REVIEW

File: pilot/pkg/security/authz/model/generator.go [hunk 1]
```diff
@@ -194,7 +194,7 @@ func (requestHeaderGenerator) principal(key, value string, forTCP bool) (*rbacpb
 	if err != nil {
 		return nil, err
 	}
-	m := matcher.HeaderMatcher(header, value)
+	m := matcher.HeaderMatcherWithRegex(header, value)
```
Comment (lei-tang) [inline pilot/pkg/security/authz/model/generator.go:197]: Thanks for the PR! Can you add a release notes to the PR describing how the PR changes the current header matching behavior, and potential impacts to the users that use the current header matching? Meanwhile, the docs of current header matching behavior need to be updated and a user guide for this feature needs to be added in follow-up PRs in istio.io/.

File: pilot/pkg/security/authz/matcher/header.go [hunk 2]
```diff
@@ -56,6 +57,40 @@ func HeaderMatcher(k, v string) *routepb.HeaderMatcher {
...
+		regex = regexp.QuoteMeta(v)
+	}
+	return &routepb.HeaderMatcher{
+		Name: k,
+		HeaderMatchSpecifier: &routepb.HeaderMatcher_SafeRegexMatch{
+			SafeRegexMatch: &matcherpb.RegexMatcher{
+				EngineType: &matcherpb.RegexMatcher_GoogleRe2{
+					GoogleRe2: &matcherpb.RegexMatcher_GoogleRE2{},
+				},
+				Regex: fmt.Sprintf("^%s$|^%s,|,%s,|,%s$", regex, regex, regex, regex),
```
Comment (howardjohn) [inline pilot/pkg/security/authz/matcher/header.go:88]: This feels like we may be introducing a high runtime cost to all users in all cases to support a niche feature - and risk of bugs

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 3]
```diff
@@ -63,6 +63,55 @@ func TestHeaderMatcher(t *testing.T) {
...
+			K:    ":path",
+			V:    "/productpage",
+			Expect: &routepb.HeaderMatcher{
+				Name: ":path",
+				HeaderMatchSpecifier: &routepb.HeaderMatcher_SafeRegexMatch{
+					SafeRegexMatch: &matcherpb.RegexMatcher{
+						EngineType: &matcherpb.RegexMatcher_GoogleRe2{
+							GoogleRe2: &matcherpb.RegexMatcher_GoogleRE2{},
+						},
+						Regex: "^/productpage$|^/productpage,|,/productpage,|,/productpage$",
```
Comment (howardjohn) [inline pilot/pkg/security/authz/matcher/header_test.go:84]: I think this introduces an expectation that commas will not have any spaces around them - not sure this is true in real world?

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 4]
```diff
@@ -63,6 +63,55 @@ func TestHeaderMatcher(t *testing.T) {
...
+			K:    ":path",
+			V:    "*/productpage*",
+			Expect: &routepb.HeaderMatcher{
+				Name: ":path",
+				HeaderMatchSpecifier: &routepb.HeaderMatcher_SafeRegexMatch{
+					SafeRegexMatch: &matcherpb.RegexMatcher{
+						EngineType: &matcherpb.RegexMatcher_GoogleRe2{
+							GoogleRe2: &matcherpb.RegexMatcher_GoogleRE2{},
+						},
+						Regex: "^.*/productpage\\*$|^.*/productpage\\*,|,.*/productpage\\*,|,.*/productpage\\*$",
```
Comment (howardjohn) [inline pilot/pkg/security/authz/matcher/header_test.go:100]: Very worried about this complexity -- if a simple rule becomes this complex regex, its hard to grok the security posture and settings, and is a big risk

**Review by howardjohn — CHANGES_REQUESTED** [review]
This seems like we are being overly lenient and will lead to an RBAC bypass.

So before:

policy: `ALLOW x-foo: bar`

Now I send `curl -H "x-foo: bar" -H "x-foo: not-bar-attack"`. The backend reads the last value only, we bypassed their policy.

**Comment by howardjohn** [comment]
IMO _if_ we  want to make this change we should add native matching support to Envoy instead of hacking around with complex regex. However, i am not sure it is safe even then (per comment above)

**Comment by edieYoung** [comment]
> This seems like we are being overly lenient and will lead to an RBAC bypass.
> So before:
> policy: ALLOW x-foo: bar
> Now I send curl -H "x-foo: bar" -H "x-foo: not-bar-attack". The backend reads the last value only, we bypassed their policy.

This header will be merged into `x-foo: bar,not-bar-attack` as per this Envoy [change](https://github.com/envoyproxy/envoy/commit/2c60632d41555ec8b3d9ef5246242be637a2db0f), and if the policy is to allow `x-foo: bar`, I think [by definition](https://istio.io/latest/docs/reference/config/security/conditions/#supported-conditions) it should be allowed. 
If user wants to match all of the header values, they should specify multiple conditions under the same rule.

**Comment by edieYoung** [comment]
> IMO _if_ we want to make this change we should add native matching support to Envoy instead of hacking around with complex regex. However, i am not sure it is safe even then (per comment above)

Agreed a native matching support to Envoy would be a better choice. Here is the [issue](https://github.com/envoyproxy/envoy/issues/13751) but it is kind of stale.

**Comment by edieYoung** [comment]
@howardjohn Thanks for your comments, I agree with your concern, that's why I am considering maybe we can make this fix an opt-in for users, so that users can opt-in/out this support by specifying a mesh config flag. They will need to make the judgement on whether or not to turn on in order to fit for their own system. wdyt?

For example, in our system, this semantics is important to our authz framework and the requests using this support is not performance sensitive(we conducted perf test and the results are acceptable given our scale).

File: pilot/pkg/security/authz/matcher/header_test.go (hunk 3 above)
Comment (edieYoung) [reply pilot/pkg/security/authz/matcher/header_test.go:84]: it should be true, with this change in Envoy https://github.com/envoyproxy/envoy/commit/2c60632d41555ec8b3d9ef5246242be637a2db0f

File: pilot/pkg/security/authz/matcher/header.go (hunk 2 above)
Comment (edieYoung) [reply pilot/pkg/security/authz/matcher/header.go:88]: maybe we can consider to make it an option for users, replied in https://github.com/istio/istio/pull/40499#issuecomment-1217215266

File: pilot/pkg/security/authz/model/generator.go (hunk 1 above)
Comment (edieYoung) [reply pilot/pkg/security/authz/model/generator.go:197]: Thanks @lei-tang, I will update it when we reach a consensus on the solution :)

Comment (lei-tang) [reply pilot/pkg/security/authz/model/generator.go:197]: Replacing HeaderMatcher() with HeaderMatcherWithRegex() changes the current behavior. Whether HeaderMatcherWithRegex() is used or not should be protected by a feature-enablement flag.

Comment (edieYoung) [reply pilot/pkg/security/authz/model/generator.go:197]: a feature-enablement flag sgtm.

File: pilot/pkg/security/authz/model/generator_test.go [hunk 5]
```diff
@@ -204,9 +204,11 @@ func TestGenerator(t *testing.T) {
 			key:   "request.headers[x-foo]",
 			value: "foo",
 			want: yamlPrincipal(t, `
-         header:
-          exactMatch: foo
-          name: x-foo`),
+        header:
```
Comment (lei-tang) [inline pilot/pkg/security/authz/model/generator_test.go:207]: The test for the original behavior should remain. The test for the new feature should be a separate test.

**Review by lei-tang — CHANGES_REQUESTED** [review]

**Comment by edieYoung** [comment]
@lei-tang @howardjohn another solution for this issue, do you think it would be more acceptable if we just expose a regex match field in the header Authz API? For example:
```
when:
      - key: request.headers[x-foo]
        regex: ""
```

**Comment by lei-tang** [comment]
> @lei-tang @howardjohn another solution for this issue, do you think it would be more acceptable if we just expose a regex match field in the header Authz API? For example:
> 
> ```
> when:
>       - key: request.headers[x-foo]
>         regex: ""
> ```

Can you add more details to this alternative solution? For example, a detailed example of how users use this alternative solution.

**Comment by edieYoung** [comment]
@lei-tang this alternative solution is to allow users to specify the regex in the authz condition, so that we can specify any regex as need, here is an example API, 

```
rules:
- when:
    - key: request.headers[x-foo]
       regex: `'^bar,.*`
```

**Comment by lei-tang** [comment]
> @lei-tang this alternative solution is to allow users to specify the regex in the authz condition, so that we can specify any regex as need, here is an example API,
> 
> ```
> rules:
> - when:
>     - key: request.headers[x-foo]
>        regex: `'^bar,.*`
> ```

There are concerns on supporting regex in authz policy, e.g., https://github.com/istio/istio/issues/16585. 
@edieYoung If you like, maybe you can write a proposal for supporting regex in authz policy and present it on Istio WG meeting to get feedback?

**Comment by rrobb-aspen** [comment]
> #39680

We also have the same concern.  We have put a PR into Istio API for this issue.

**Comment by edieYoung** [comment]
Hey @lei-tang @howardjohn, circling back on this, **being able to define a policy rule to match among multiple values in the inline header as per this Envoy's [change](https://github.com/envoyproxy/envoy/issues/13751)** is still a demanding feature for many of us (Note that the ask is not to support arbitrary regex or contain match, etc, they are just some of the solutions), and we have been using this patch in our fork for a while and lived happy with it . 

To continue on this, consider the concern of breaking the existing behavior, what about introducing a separate condition attribute called `request.inline-headers` to distinguish from the original header attribute? By using this attribute, we can clearly describe how it behaves differently from `request.header` that it can match the value among the multiple values in inline header format while `request.header` does not. So that only the users who explicitly know they are expecting inline headers will consciously use this feature.

The new usage would be like:
The following policy will allow the request with header in the inline format `x-foo: bar,baz`:
```
ALLOW
rules:
- when:
    - key: request.ineline-headers[x-foo]
       value: bar
```

I will send out a PR once it's ready, but just want to get some quick feedbacks, also please advise any process I need to go through to introduce this new feature, thanks in advance!

**Comment by howardjohn** [comment]
@edieYoung have we looked into adding native support in Envoy? Multi header sematnics is very very complex - I don't feel comfortable implementing it with a regex regardless of the user facing API

**Comment by edieYoung** [comment]
@howardjohn totally agree this semantics is complex, but I think it's worthy to do something to catch up given inline header format is the standard as well.
Unfortunately, Envoy doesn't add support for this specifically, it provides multiple matching types to leave the choice of semantics to user-side implementation. (reading from https://github.com/envoyproxy/envoy/issues/13751, the proposed solution is only about providing generic matching API, which provides contains/regex match in string match by treating header value merely as a string).

In my opinion, the use of Regex in this case is not a major concern since we are only using a fixed and strict regex as an internal implementation detail and restricting its blast radius through an opt-in API. To be extra cautious, we could label it as experimental, such as `request.experimental-inline-header`. What do you think?

This is the draft PR https://github.com/istio/istio/pull/46183.

**Comment by howardjohn** [comment]
I really don't believe regex will ever be acceptable here. Aside from performance concerns, I don't think its secure or correct. See https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.1.2 - its not so simple. `foo , ,bar,charlie` is a valid header.

Can you parse a lot of headers with regex this way - sure... but this is a security policy. We need to be 100% certain that we can parse **all** headers correctly, or we have a CVE. I am doubtful I will ever have that level of confidence with a regex.

**Comment by edieYoung** [comment]
I see, I guess the most ideal way would be having Envoy to support something like ListMatcher which can parse the header value as a list properly.
However, in current situation, multiple value headers will always be concatenated with comma by Envoy(https://github.com/envoyproxy/envoy/commit/2c60632d41555ec8b3d9ef5246242be637a2db0f) before matching, so it's 100% that we are always expecting a comma separated string(no blank space, no empty value).

**Comment by howardjohn** [comment]
> so it's 100% that we are always expecting a comma separated string(no blank space, no empty value).

I don't think this is true... `curl echo -H "foo: bar, , baz"` will very likely have spaces, for example. And that is just one case. I have no confidence at all that Istio (especially with regex) is going to properly handle all the crazy edge cases of HTTP. Even envoy did not without regex, hence the CVE you linked.

**Comment by karloramone** [comment]
Is there any plans to continue with this PR?

**Comment by edieYoung** [comment]
we'd love to continue this if we can get approvals from the community, since this is stale, I just revived in another PR https://github.com/istio/istio/pull/56085 and made it a separate attribute so the original header attribute doesn't change.

**Pull request closed without merging**